src/common/scram.c

```diff
@@ -0,0 +1,297 @@
+	                                        data, client_final_message_without_proof);
+
+	// ClientKey := HMAC(SaltedPassword, "Client Key")
+	client_key = g_malloc0(session->digest_size);
+
+	HMAC(session->digest, session->salted_password, session->digest_size,
+	     (unsigned char *) CLIENT_KEY, strlen(CLIENT_KEY), client_key, &client_key_len);
+
+	// StoredKey := H(ClientKey)
+	if (!create_SHA(session, client_key, session->digest_size, stored_key, &stored_key_len)) {
+		return SCRAM_ERROR;
+	}
+
+	// ClientSignature := HMAC(StoredKey, AuthMessage)
+	client_signature = g_malloc0(session->digest_size);
+	HMAC(session->digest, stored_key, stored_key_len, (unsigned char *) session->auth_message,
+	     strlen((char *) session->auth_message), client_signature, NULL);
+
+	// ClientProof := ClientKey XOR ClientSignature
+	client_proof = g_malloc0(client_key_len);
+
+	for (i = 0; i < client_key_len; i++) {
+		client_proof[i] = client_key[i] ^ client_signature[i];
+	}
+
+	client_proof_b64 = g_base64_encode((guchar *) client_proof, client_key_len);
+
+	*output = g_strdup_printf("%s,p=%s", client_final_message_without_proof, client_proof_b64);
+	*output_len = strlen(*output);
+
+	g_free(server_nonce_b64);
+	g_free(client_final_message_without_proof);
+	g_free(salt);
+	g_free(client_signature);
+	g_free(client_proof);
+
+	session->step++;
+	return SCRAM_IN_PROGRESS;
+}
+
+static int process_server_final(scram_session *session, const char *data)
+{
+	char *verifier;
+	unsigned char *server_key, *server_signature;
+	unsigned int verifier_len = 0, server_key_len = 0, server_signature_len = 0;
+	gint b64_state = 0;
+	guint b64_save = 0;
+
+	if (strlen(data) < 3 || (data[0] != 'v' && data[1] != '=')) {
+		return SCRAM_ERROR;
+	}
+
+	verifier = g_strdup(data + 2);
+	verifier_len = g_base64_decode_step(verifier, strlen(verifier), (guchar *) verifier,
+	                                    &b64_state, &b64_save);
+
+	// ServerKey := HMAC(SaltedPassword, "Server Key")
+	server_key = g_malloc0(session->digest_size);
+	HMAC(session->digest, session->salted_password, session->digest_size,
+	     (unsigned char *) SERVER_KEY, strlen(SERVER_KEY), server_key, &server_key_len);
+
+	// ServerSignature := HMAC(ServerKey, AuthMessage)
+	server_signature = g_malloc0(session->digest_size);
+	HMAC(session->digest, server_key, session->digest_size,
+	     (unsigned char *) session->auth_message, strlen((char *) session->auth_message),
+	     server_signature, &server_signature_len);
+
+	if (verifier_len == server_signature_len &&
+	    memcmp(verifier, server_signature, verifier_len) == 0) {
+		g_free(verifier);
+		g_free(server_key);
+		g_free(server_signature);
+		return SCRAM_SUCCESS;
+	} else {
+		g_free(verifier);
+		g_free(server_key);
+		g_free(server_signature);
+		return SCRAM_ERROR;
+	}
+}
+
+int scram_process(scram_session *session, const char *input, char **output, size_t *output_len)
+{
+	int ret = 0;
+
+	switch (session->step) {
+	case 0:

```

Comment:
It would be nice to turn the steps into an enum with readable names.

Reply:
This is more a step counter, so int makes sense imho. Also it is only evaluated in scram_process().

0 = nothing done so far, send client-first-message
1 = parse server-first-message and send client-final-message
2 = parse/verify server-final-message

Reply:
Ok, not a big deal.